### PR TITLE
[SPARK-32978][SQL] Make sure the number of dynamic part metric is correct

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?

The purpose of this pr is to resolve SPARK-32978.

The main reason of bad case describe in SPARK-32978 is the `BasicWriteTaskStatsTracker` directly reports the new added partition number of each task, which makes it impossible to remove duplicate data in driver side.

The main of this pr is change to report partitionValues to driver and remove duplicate data at driver side to make sure the number of dynamic part metric is correct.
 
### Why are the changes needed?
The the number of dynamic part metric we display on the UI should be correct.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add a new test case refer to described in SPARK-32978

### CODE REVIEW

**Comment by LuciferYang** [comment]
cc @wangyum , the current result should be correct, but will increase memory and network pressure

**Comment by wangyum** [comment]
cc @cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/BasicWriteJobStatsTrackerMetricSuite.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      // There should be 4 SQLExecutionUIData in executionsList and the 3rd item is we need,
+      // but the executionId is indeterminate in maven test,
+      // so the `statusStore.execution(executionId)` API is not used.
+      assert(statusStore.executionsCount() == 4)
+      val executionData = statusStore.executionsList()(2)
+      val accumulatorIdOpt =
+        executionData.metrics.find(_.name == "number of dynamic part").map(_.accumulatorId)
+      assert(accumulatorIdOpt.isDefined)
+      val numPartsOpt = executionData.metricValues.get(accumulatorIdOpt.get)
+      assert(numPartsOpt.isDefined && numPartsOpt.get == partitions)
```
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/BasicWriteJobStatsTrackerMetricSuite.scala:52]: Without change of `BasicWriteStatsTracker.scala`, the numParts will be 200

**Comment by AngersZhuuuu** [comment]
I have thought to do like this too, but if partition num is too big, return partition value will occupy too many memory.

**Comment by LuciferYang** [comment]
@AngersZhuuuu That's what I'm worried about :(

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala [hunk 2]
```diff
@@ -30,12 +32,13 @@ import org.apache.spark.sql.execution.metric.{SQLMetric, SQLMetrics}
...
 
 
+
 /**
  * Simple metrics collected during an instance of [[FileFormatDataWriter]].
  * These were first introduced in https://github.com/apache/spark/pull/18159 (SPARK-20703).
  */
 case class BasicWriteTaskStats(
-    numPartitions: Int,
+    partitions: Seq[InternalRow],
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:41]: This increases the data size we need to transfer between executors and the driver. Do we have a microbenchmark to verify the impact?

Comment (LuciferYang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:41]: A benchmark for latency or memory usage? I can try it ~

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:41]: end-to-end performance of an INSERT query, with a partitioned table with 1 or 2 or 3 partitioned columns.

Comment (LuciferYang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:41]: ok ~ busy with my own work, will give feedback later

**Comment by LuciferYang** [comment]
Address 724eee6 add a simple microbenchmark/

**With this pr** the result is :

```
Running benchmark: dynamic insert table benchmark
  Running case: insert table with 10000 rows, one partition column, 10 partitions
  Stopped after 2 iterations, 2134 ms
  Running case: insert table with 10000 rows, one partition column, 50 partitions
  Stopped after 2 iterations, 7206 ms
  Running case: insert table with 10000 rows, one partition column, 100 partitions
  Stopped after 2 iterations, 9105 ms
  Running case: insert table with 10000 rows, one partition column, 200 partitions
  Stopped after 2 iterations, 14778 ms
  Running case: insert table with 10000 rows, one partition column, 500 partitions
  Stopped after 2 iterations, 42992 ms
  Running case: insert table with 10000 rows, two partition columns, 10 partitions
  Stopped after 2 iterations, 2331 ms
  Running case: insert table with 10000 rows, two partition columns, 50 partitions
  Stopped after 2 iterations, 6768 ms
  Running case: insert table with 10000 rows, two partition columns, 100 partitions
  Stopped after 2 iterations, 9274 ms
  Running case: insert table with 10000 rows, two partition columns, 200 partitions
  Stopped after 2 iterations, 17487 ms
  Running case: insert table with 10000 rows, two partition columns, 500 partitions
  Stopped after 2 iterations, 54044 ms
  Running case: insert table with 10000 rows, three partition columns, 10 partitions
  Stopped after 2 iterations, 2368 ms
  Running case: insert table with 10000 rows, three partition columns, 50 partitions
  Stopped after 2 iterations, 5538 ms
  Running case: insert table with 10000 rows, three partition columns, 100 partitions
  Stopped after 2 iterations, 11687 ms
  Running case: insert table with 10000 rows, three partition columns, 200 partitions
  Stopped after 2 iterations, 22371 ms
  Running case: insert table with 10000 rows, three partition columns, 500 partitions
  Stopped after 2 iterations, 55828 ms

Java HotSpot(TM) 64-Bit Server VM 1.8.0_192-b12 on Mac OS X 10.15.7
Intel(R) Core(TM) i5-7360U CPU @ 2.30GHz
dynamic insert table benchmark:                                        Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
-----------------------------------------------------------------------------------------------------------------------------------------------------
insert table with 10000 rows, one partition column, 10 partitions                922           1067         206          0.0       92182.2       1.0X
insert table with 10000 rows, one partition column, 50 partitions               3265           3603         478          0.0      326535.4       0.3X
insert table with 10000 rows, one partition column, 100 partitions              4390           4553         230          0.0      438988.3       0.2X
insert table with 10000 rows, one partition column, 200 partitions              6585           7389        1137          0.0      658477.7       0.1X
insert table with 10000 rows, one partition column, 500 partitions             20220          21496        1805          0.0     2022011.5       0.0X
insert table with 10000 rows, two partition columns, 10 partitions              1114           1166          72          0.0      111432.2       0.8X
insert table with 10000 rows, two partition columns, 50 partitions              2467           3384        1297          0.0      246670.3       0.4X
insert table with 10000 rows, two partition columns, 100 partitions             4559           4637         110          0.0      455904.3       0.2X
insert table with 10000 rows, two partition columns, 200 partitions             8631           8744         159          0.0      863130.8       0.1X
insert table with 10000 rows, two partition columns, 500 partitions            23806          27022        1498          0.0     2380574.6       0.0X
insert table with 10000 rows, three partition columns, 10 partitions            1096           1184         125          0.0      109639.4       0.8X
insert table with 10000 rows, three partition columns, 50 partitions            2694           2769         107          0.0      269364.4       0.3X
insert table with 10000 rows, three partition columns, 100 partitions           5701           5844         202          0.0      570137.3       0.2X
insert table with 10000 rows, three partition columns, 200 partitions          11105          11186         115          0.0     1110452.3       0.1X
insert table with 10000 rows, three partition columns, 500 partitions          26978          27914        1324          0.0     2697786.6       0.0X
```

**Without this pr** the result is :

```
Running benchmark: dynamic insert table benchmark
  Running case: insert table with 10000 rows, one partition column, 10 partitions
  Stopped after 3 iterations, 2356 ms
  Running case: insert table with 10000 rows, one partition column, 50 partitions
  Stopped after 2 iterations, 6328 ms
  Running case: insert table with 10000 rows, one partition column, 100 partitions
  Stopped after 2 iterations, 8942 ms
  Running case: insert table with 10000 rows, one partition column, 200 partitions
  Stopped after 2 iterations, 17401 ms
  Running case: insert table with 10000 rows, one partition column, 500 partitions
  Stopped after 2 iterations, 43009 ms
  Running case: insert table with 10000 rows, two partition columns, 10 partitions
  Stopped after 3 iterations, 2024 ms
  Running case: insert table with 10000 rows, two partition columns, 50 partitions
  Stopped after 2 iterations, 4862 ms
  Running case: insert table with 10000 rows, two partition columns, 100 partitions
  Stopped after 2 iterations, 11229 ms
  Running case: insert table with 10000 rows, two partition columns, 200 partitions
  Stopped after 2 iterations, 18244 ms
  Running case: insert table with 10000 rows, two partition columns, 500 partitions
  Stopped after 2 iterations, 54922 ms
  Running case: insert table with 10000 rows, three partition columns, 10 partitions
  Stopped after 3 iterations, 2173 ms
  Running case: insert table with 10000 rows, three partition columns, 50 partitions
  Stopped after 2 iterations, 5660 ms
  Running case: insert table with 10000 rows, three partition columns, 100 partitions
  Stopped after 2 iterations, 14925 ms
  Running case: insert table with 10000 rows, three partition columns, 200 partitions
  Stopped after 2 iterations, 28378 ms
  Running case: insert table with 10000 rows, three partition columns, 500 partitions
  Stopped after 2 iterations, 59941 ms

Java HotSpot(TM) 64-Bit Server VM 1.8.0_192-b12 on Mac OS X 10.15.7
Intel(R) Core(TM) i5-7360U CPU @ 2.30GHz
dynamic insert table benchmark:                                        Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
-----------------------------------------------------------------------------------------------------------------------------------------------------
insert table with 10000 rows, one partition column, 10 partitions                715            785          98          0.0       71450.9       1.0X
insert table with 10000 rows, one partition column, 50 partitions               3132           3164          46          0.0      313201.2       0.2X
insert table with 10000 rows, one partition column, 100 partitions              4375           4471         136          0.0      437546.1       0.2X
insert table with 10000 rows, one partition column, 200 partitions              7968           8701        1035          0.0      796846.9       0.1X
insert table with 10000 rows, one partition column, 500 partitions             19208          21505         NaN          0.0     1920778.4       0.0X
insert table with 10000 rows, two partition columns, 10 partitions               600            675          66          0.0       60027.5       1.2X
insert table with 10000 rows, two partition columns, 50 partitions              2372           2431          83          0.0      237244.2       0.3X
insert table with 10000 rows, two partition columns, 100 partitions             5385           5615         325          0.0      538471.6       0.1X
insert table with 10000 rows, two partition columns, 200 partitions             8496           9122         885          0.0      849591.8       0.1X
insert table with 10000 rows, two partition columns, 500 partitions            25747          27461        2424          0.0     2574722.7       0.0X
insert table with 10000 rows, three partition columns, 10 partitions             687            725          35          0.0       68748.8       1.0X
insert table with 10000 rows, three partition columns, 50 partitions            2757           2830         104          0.0      275692.0       0.3X
insert table with 10000 rows, three partition columns, 100 partitions           6336           7463        1594          0.0      633568.3       0.1X
insert table with 10000 rows, three partition columns, 200 partitions          14046          14189         202          0.0     1404645.4       0.1X
insert table with 10000 rows, three partition columns, 500 partitions          26749          29971        1520          0.0     2674929.0       0.0X

```


cc @cloud-fan seems no essential difference

**Comment by AngersZhuuuu** [comment]
return size is partition num * shuffle num always can be millions level

**Comment by cloud-fan** [comment]
> return size is partition num * shuffle num always can be millions level

I thought about it. If a table has 10k partitions, it's unlikely that each write task touches all the 10k partitions. So the total size is not that large.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala [hunk 3]
```diff
@@ -30,12 +32,13 @@ import org.apache.spark.sql.execution.metric.{SQLMetric, SQLMetrics}
 import org.apache.spark.util.SerializableConfiguration
 
 
+
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:35]: unnecessary change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala [hunk 4]
```diff
@@ -76,7 +79,7 @@ class BasicWriteTaskStatsTracker(hadoopConf: Configuration)
 
 
   override def newPartition(partitionValues: InternalRow): Unit = {
-    numPartitions += 1
+    partitions = partitions :+ partitionValues
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:82]: this looks like appending a immutable collection. Can we be more explicit? e.g. `partitions.append(partitionValues)`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala [hunk 5]
```diff
@@ -139,20 +142,22 @@ class BasicWriteJobStatsTracker(
...
+    var partitionsSet: mutable.Set[InternalRow] = mutable.HashSet.empty
     var numFiles: Long = 0L
     var totalNumBytes: Long = 0L
     var totalNumOutput: Long = 0L
 
     val basicStats = stats.map(_.asInstanceOf[BasicWriteTaskStats])
 
     basicStats.foreach { summary =>
-      numPartitions += summary.numPartitions
+      partitionsSet ++= summary.partitions
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:153]: ditto, `partitionsSet.addAll(summary.partitions)`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala [hunk 6]
```diff
@@ -139,20 +142,22 @@ class BasicWriteJobStatsTracker(
...
       numFiles += summary.numFiles
       totalNumBytes += summary.numBytes
       totalNumOutput += summary.numRows
     }
 
+    val numPartitions: Long = partitionsSet.size
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:159]: nit: it's only used once, we can inline it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala (hunk 4 above)
Comment (LuciferYang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:82]: `partitions.appended(partitionValues)` need Scala 2.13

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala (hunk 5 above)
Comment (LuciferYang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:153]: ditto, `partitionsSet.addAll(summary.partitions)` can only be used in Scala 2.13 too.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala (hunk 6 above)
Comment (LuciferYang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala]: Address 5769222 fix this

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:81]: or `add`? doesn't scala 2.12 have java-like method to add elements?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+      part1Step: Int, part2Step: Int, part3Step: Int): Long = {
+    val dataFrame = spark.range(0, numberRows, 1, 4)
+    val dataFrame1 = spark.range(0, numberRows, part1Step, 4)
+    val dataFrame2 = spark.range(0, numberRows, part2Step, 4)
+    val dataFrame3 = spark.range(0, numberRows, part3Step, 4)
+
+    val data = dataFrame.join(dataFrame1).join(dataFrame2).join(dataFrame3)
+      .toDF("id", "part1", "part2", "part3")
+
+    data.createOrReplaceTempView("tmpTable")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala:50]: `tmpTable` -> `tmpView`

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala:50]: BTW is it the same as `data.write.saveAsTable("sourceTable")`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        s"part1, " +
+        s"part2, " +
+        s"part3 " +
+        s"from sourceTable")
+    }
+  }
+
+  override def runBenchmarkSuite(mainArgs: Array[String]): Unit = {
+    val sourceTable = "sourceTable"
+    val tableOnePartitionColumn = "tableOnePartitionColumn"
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala:114]: it can be simpler `val onePartTable =  ...`

**Comment by LuciferYang** [comment]
Address 73e2ea6 reorganize the benchmark file

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala (hunk 4 above)
Comment (LuciferYang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala]: Address 6d80788 change to use `ArrayBuffer` to store `partitions ` and there change to `partitions.appended(partitionValues)`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+      "using parquet partitioned by (part1, part2)")
+    spark.sql("create table " +
+      "tableThreePartitionColumn(i bigint, part1 bigint, part2 bigint, part3 bigint) " +
+      "using parquet partitioned by (part1, part2, part3)")
+  }
+
+  def writeOnePartitionColumnTable(partitionNumber: Long, benchmark: Benchmark): Unit = {
+    benchmark.addCase(s"one partition column, $partitionNumber partitions") { _ =>
+      spark.sql("insert overwrite table " +
+        "tableOnePartitionColumn partition(part) " +
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala]: done~

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala (hunk 8 above)
Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala]: done ~

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala (hunk 7 above)
Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala]: done ~

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    data.count()
+  }
+
+  def writeOnePartitionColumnTable(tableName: String,
+      partitionNumber: Long, benchmark: Benchmark): Unit = {
+    spark.sql(s"create table $tableName(i bigint, part bigint) " +
+      "using parquet partitioned by (part)")
+    benchmark.addCase(s"one partition column, $partitionNumber partitions") { _ =>
+      spark.sql("insert overwrite table " +
+        s"$tableName partition(part) " +
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala:59]: nit: this can be merged into the previous line

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    }
+  }
+
+  def writeTwoPartitionColumnTable(tableName: String,
+      partitionNumber: Long, benchmark: Benchmark): Unit = {
+    spark.sql(s"create table $tableName(i bigint, part1 bigint, part2 bigint) " +
+      "using parquet partitioned by (part1, part2)")
+    benchmark.addCase(s"two partition columns, $partitionNumber partitions") { _ =>
+      spark.sql("insert overwrite table " +
+        s"$tableName partition(part1, part2) " +
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala:70]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+    }
+  }
+
+  def writeThreePartitionColumnTable(tableName: String,
+      partitionNumber: Long, benchmark: Benchmark): Unit = {
+    spark.sql(s"create table $tableName(i bigint, part1 bigint, part2 bigint, part3 bigint) " +
+      "using parquet partitioned by (part1, part2, part3)")
+    benchmark.addCase(s"three partition columns, $partitionNumber partitions") { _ =>
+      spark.sql("insert overwrite table " +
+        s"$tableName partition(part1, part2, part3) " +
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala:81]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/BasicWriteJobStatsTrackerMetricSuite.scala [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    try {
+      val partitions = "50"
+      spark = SparkSession.builder().master("local[4]").getOrCreate()
+      val statusStore = spark.sharedState.statusStore
+      val oldExecutionsSize = statusStore.executionsList().size
+
+      spark.sql("create table dynamic_partition(i bigint, part bigint) " +
+        "using parquet partitioned by (part)").collect()
+      spark.sql("insert overwrite table dynamic_partition partition(part) " +
+        s"select id, id % $partitions as part from range(10000)").collect()
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/BasicWriteJobStatsTrackerMetricSuite.scala:35]: nit: we don't need to call `.collect` to trigger DDL/DML commands. `sql(...)` already does the job.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/BasicWriteJobStatsTrackerMetricSuite.scala [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+      // Wait for listener to finish computing the metrics for the executions.
+      while (statusStore.executionsList().size - oldExecutionsSize < 4 ||
+        statusStore.executionsList().last.metricValues == null) {
+        Thread.sleep(100)
+      }
+
+      // There should be 4 SQLExecutionUIData in executionsList and the 3rd item is we need,
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/BasicWriteJobStatsTrackerMetricSuite.scala:43]: why there are 4? is it because of `collect`?

BTW can we call `val oldExecutionsSize = statusStore.executionsList().size` after create table? then we just need to wait for one `SQLExecutionUIData`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala (hunk 10 above)
Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala]: done

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala (hunk 11 above)
Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala]: done

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala (hunk 12 above)
Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala]: done

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/BasicWriteJobStatsTrackerMetricSuite.scala (hunk 13 above)
Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/BasicWriteJobStatsTrackerMetricSuite.scala:35]: some other problems without call `.collect`, let me re-check this.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/BasicWriteJobStatsTrackerMetricSuite.scala (hunk 14 above)
Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/BasicWriteJobStatsTrackerMetricSuite.scala]: Address 15c7519 fix this

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/BasicWriteJobStatsTrackerMetricSuite.scala (hunk 13 above)
Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/BasicWriteJobStatsTrackerMetricSuite.scala]: no other problems, Address 15c7519 fix this

**Comment by LuciferYang** [comment]
cc @cloud-fan Any other problems need to be fixed?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+import org.apache.spark.benchmark.Benchmark
+import org.apache.spark.sql.SparkSession
+
+/**
+ * Benchmark to measure insert into table with dynamic partition columns.
+ * To run this benchmark:
+ * {{{
+ *   1. without sbt: bin/spark-submit --class <this class> <spark sql test jar>
+ *   2. build/sbt "sql/test:runMain <this class>"
+ *   3. generate result: SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "sql/test:runMain <this class>"
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala:29]: We need to run it and commit the benchmark result to the codebase.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+ *      Results will be written to
+ *      "benchmarks/InsertTableWithDynamicPartitionsBenchmark-results.txt".
+ * }}}
+ */
+object InsertTableWithDynamicPartitionsBenchmark extends DataSourceWriteBenchmark {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala:34]: Seems we don't use the utils methods in `DataSourceWriteBenchmark`. I think we can implement `SqlBasedBenchmark` directly.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+
+  override def getSparkSession: SparkSession = {
+    SparkSession.builder().master("local[4]").getOrCreate()
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala:37]: do we have to override it? The default one-thread spark session is better to reason about benchmark results.

**Review by cloud-fan — APPROVED** [review]
LGTM except minor comments to the benchmark

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala (hunk 16 above)
Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala:34]: `withTable` in `DataSourceWriteBenchmark ` is used to clean up the created table resources after benchmark

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala (hunk 17 above)
Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala]: Address fa01951 fix this

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala (hunk 15 above)
Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/InsertTableWithDynamicPartitionsBenchmark.scala:28]: Add 4f47e20 upload benchmark result file

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by LuciferYang** [comment]
thx~ @cloud-fan @wangyum @AngersZhuuuu
